NPE when parsing ScopeId in some cases

When you miss configure the `shiro.ini` file in the `rest-api-web` module and the session is not added for a resource, a NPE is thrown when parsing `ScopeId` path param.

When is configured properly this does not happen, but it makes sense to handle this situation and trhowing a proper error explaining the cause.

Stacktrace:

> 2:10:17,346 [http-nio-8090-exec-2] ERROR ThrowableMapper:30  - Severe error!
> org.glassfish.jersey.server.ParamException$PathParamException: HTTP 404 Not Found
> 	at org.glassfish.jersey.server.internal.inject.PathParamValueFactoryProvider$PathParamValueFactory.provide(PathParamValueFactoryProvider.java:95)
> 	at org.glassfish.jersey.server.spi.internal.ParamValueFactoryWithSource.provide(ParamValueFactoryWithSource.java:71)
> 	at org.glassfish.jersey.server.spi.internal.ParameterValueHelper.getParameterValues(ParameterValueHelper.java:90)
> 	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$AbstractMethodParamInvoker.getParamValues(JavaResourceMethodDispatcherProvider.java:127)
> 	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:205)
> 	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)
> 	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)
> 	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)
> 	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)
> 	at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326)
> 	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
> 	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
> 	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
> 	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
> 	at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
> 	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
> 	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305)
> 	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154)
> 	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473)
> 	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427)
> 	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388)
> 	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341)
> 	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228)
> 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:292)
> 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
> 	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
> 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240)
> 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
> 	at org.apache.shiro.web.servlet.AbstractShiroFilter.executeChain(AbstractShiroFilter.java:449)
> 	at org.apache.shiro.web.servlet.AbstractShiroFilter$1.call(AbstractShiroFilter.java:365)
> 	at org.apache.shiro.subject.support.SubjectCallable.doCall(SubjectCallable.java:90)
> 	at org.apache.shiro.subject.support.SubjectCallable.call(SubjectCallable.java:83)
> 	at org.apache.shiro.subject.support.DelegatingSubject.execute(DelegatingSubject.java:383)
> 	at org.apache.shiro.web.servlet.AbstractShiroFilter.doFilterInternal(AbstractShiroFilter.java:362)
> 	at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
> 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240)
> 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
> 	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:212)
> 	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
> 	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:502)
> 	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:141)
> 	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
> 	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:616)
> 	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
> 	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:509)
> 	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1104)
> 	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:684)
> 	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1520)
> 	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1476)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
> 	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
> 	at java.lang.Thread.run(Thread.java:745)
> Caused by: org.glassfish.jersey.internal.inject.ExtractorException: java.lang.NullPointerException
> 	at org.glassfish.jersey.server.internal.inject.ParamConverters$AbstractStringReader.fromString(ParamConverters.java:95)
> 	at org.glassfish.jersey.server.internal.inject.AbstractParamValueExtractor.convert(AbstractParamValueExtractor.java:139)
> 	at org.glassfish.jersey.server.internal.inject.AbstractParamValueExtractor.fromString(AbstractParamValueExtractor.java:130)
> 	at org.glassfish.jersey.server.internal.inject.SingleValueExtractor.extract(SingleValueExtractor.java:84)
> 	at org.glassfish.jersey.server.internal.inject.PathParamValueFactoryProvider$PathParamValueFactory.provide(PathParamValueFactoryProvider.java:93)
> 	... 52 more
> Caused by: java.lang.NullPointerException
> 	at org.eclipse.kapua.app.api.resources.v1.resources.model.ScopeId.<init>(ScopeId.java:52)
> 	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
> 	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
> 	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
> 	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
> 	at org.glassfish.jersey.server.internal.inject.ParamConverters$StringConstructor$1._fromString(ParamConverters.java:132)
> 	at org.glassfish.jersey.server.internal.inject.ParamConverters$AbstractStringReader.fromString(ParamConverters.java:85)
> 	... 56 more
>

--- a/rest-api/resources/src/main/java/org/eclipse/kapua/app/api/resources/v1/resources/exception/RestApiErrorCodes.java
+++ b/rest-api/resources/src/main/java/org/eclipse/kapua/app/api/resources/v1/resources/exception/RestApiErrorCodes.java
@@ -1,0 +1,22 @@
+/*******************************************************************************
+ * Copyright (c) 2018 Eurotech and/or its affiliates and others
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Eurotech - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.kapua.app.api.resources.v1.resources.exception;
+
+import org.eclipse.kapua.KapuaErrorCode;
+
+public enum RestApiErrorCodes implements KapuaErrorCode {
+    /**
+     * When a resource receive a request, but the {@link org.eclipse.kapua.commons.security.KapuaSession} is not populated.
+     * This is likely to happen when the `shiro.ini` does not map the `kapuaAuthcAccessToken` as a request filter chain.
+     */
+    SESSION_NOT_POPULATED
+}

--- a/rest-api/resources/src/main/java/org/eclipse/kapua/app/api/resources/v1/resources/exception/RestApiRuntimeException.java
+++ b/rest-api/resources/src/main/java/org/eclipse/kapua/app/api/resources/v1/resources/exception/RestApiRuntimeException.java
@@ -1,0 +1,61 @@
+/*******************************************************************************
+ * Copyright (c) 2018 Eurotech and/or its affiliates and others
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Eurotech - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.kapua.app.api.resources.v1.resources.exception;
+
+import org.eclipse.kapua.KapuaRuntimeException;
+
+public class RestApiRuntimeException extends KapuaRuntimeException {
+
+    /**
+     * Constructor
+     *
+     * @param message
+     * @param throwable
+     */
+    private RestApiRuntimeException(String message, Throwable throwable) {
+        super(message, throwable);
+    }
+
+    /**
+     * Builds a new {@link RestApiRuntimeException} instance based on the supplied {@link RestApiErrorCodes}.
+     *
+     * @param code
+     */
+    public RestApiRuntimeException(RestApiErrorCodes code) {
+        this(code, (Object[]) null);
+    }
+
+    /**
+     * Builds a new {@link RestApiRuntimeException} instance based on the supplied {@link RestApiErrorCodes}
+     * and optional arguments for the associated exception message.
+     *
+     * @param code
+     * @param arguments
+     */
+    public RestApiRuntimeException(RestApiErrorCodes code, Object... arguments) {
+        this(code, null, arguments);
+    }
+
+    /**
+     * Builds a new {@link KapuaRuntimeException} instance based on the supplied {@link RestApiErrorCodes},
+     * an Throwable cause, and optional arguments for the associated exception message.
+     *
+     * @param code
+     * @param cause
+     * @param arguments
+     */
+    public RestApiRuntimeException(RestApiErrorCodes code, Throwable cause, Object... arguments) {
+        super(cause);
+        this.code = code;
+        args = arguments;
+    }
+}

--- a/rest-api/resources/src/main/java/org/eclipse/kapua/app/api/resources/v1/resources/exception/SessionNotPopulatedException.java
+++ b/rest-api/resources/src/main/java/org/eclipse/kapua/app/api/resources/v1/resources/exception/SessionNotPopulatedException.java
@@ -1,0 +1,19 @@
+/*******************************************************************************
+ * Copyright (c) 2018 Eurotech and/or its affiliates and others
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Eurotech - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.kapua.app.api.resources.v1.resources.exception;
+
+public class SessionNotPopulatedException extends RestApiRuntimeException {
+
+    public SessionNotPopulatedException() {
+        super(RestApiErrorCodes.SESSION_NOT_POPULATED);
+    }
+}

--- a/rest-api/resources/src/main/java/org/eclipse/kapua/app/api/resources/v1/resources/model/ScopeId.java
+++ b/rest-api/resources/src/main/java/org/eclipse/kapua/app/api/resources/v1/resources/model/ScopeId.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2011, 2017 Eurotech and/or its affiliates and others
+ * Copyright (c) 2011, 2018 Eurotech and/or its affiliates and others
  *
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
@@ -11,16 +11,16 @@
  *******************************************************************************/
 package org.eclipse.kapua.app.api.resources.v1.resources.model;
 
-import java.math.BigInteger;
-import java.util.Base64;
-
-import javax.ws.rs.PathParam;
-
 import org.eclipse.kapua.app.api.core.settings.KapuaApiSetting;
 import org.eclipse.kapua.app.api.core.settings.KapuaApiSettingKeys;
+import org.eclipse.kapua.app.api.resources.v1.resources.exception.SessionNotPopulatedException;
 import org.eclipse.kapua.commons.security.KapuaSecurityUtils;
 import org.eclipse.kapua.commons.security.KapuaSession;
 import org.eclipse.kapua.model.id.KapuaId;
+
+import javax.ws.rs.PathParam;
+import java.math.BigInteger;
+import java.util.Base64;
 
 /**
  * {@link KapuaId} implementation to be used on REST API to parse the {@link PathParam} scopeId.
@@ -49,6 +49,11 @@ public class ScopeId implements KapuaId {
 
         if (SCOPE_ID_WILDCARD.equals(compactScopeId)) {
             KapuaSession session = KapuaSecurityUtils.getSession();
+
+            if (session == null) {
+                throw new SessionNotPopulatedException();
+            }
+
             setId(session.getScopeId().getId());
         } else {
             byte[] bytes = Base64.getUrlDecoder().decode(compactScopeId);

--- a/service/api/src/main/java/org/eclipse/kapua/KapuaRuntimeException.java
+++ b/service/api/src/main/java/org/eclipse/kapua/KapuaRuntimeException.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2011, 2016 Eurotech and/or its affiliates and others
+ * Copyright (c) 2011, 2018 Eurotech and/or its affiliates and others
  *
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
@@ -11,20 +11,19 @@
  *******************************************************************************/
 package org.eclipse.kapua;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import java.text.MessageFormat;
 import java.util.Locale;
 import java.util.MissingResourceException;
 import java.util.ResourceBundle;
 import java.util.StringJoiner;
 
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
 /**
  * Kapua runtime exception.
  *
  * @since 1.0
- *
  */
 public class KapuaRuntimeException extends RuntimeException {
 
@@ -41,7 +40,6 @@ public class KapuaRuntimeException extends RuntimeException {
     /**
      * Constructor
      */
-    @SuppressWarnings("unused")
     private KapuaRuntimeException() {
         super();
     }
@@ -51,8 +49,7 @@ public class KapuaRuntimeException extends RuntimeException {
      *
      * @param message
      */
-    @SuppressWarnings("unused")
-    private KapuaRuntimeException(String message) {
+    protected KapuaRuntimeException(String message) {
         this(message, null);
     }
 
@@ -61,8 +58,7 @@ public class KapuaRuntimeException extends RuntimeException {
      *
      * @param throwable
      */
-    @SuppressWarnings("unused")
-    private KapuaRuntimeException(Throwable throwable) {
+    protected KapuaRuntimeException(Throwable throwable) {
         this(null, throwable);
     }
 
@@ -72,7 +68,7 @@ public class KapuaRuntimeException extends RuntimeException {
      * @param message
      * @param throwable
      */
-    private KapuaRuntimeException(String message, Throwable throwable) {
+    protected KapuaRuntimeException(String message, Throwable throwable) {
         super(message, throwable);
     }
 

